### ISSUE TITLE
Create email categories for unsubscribes 

### ISSUE THREAD

**Issue by depietrodeanna** [episphere/connect#720]
Hi Jessica, the following categories are the ones we'll want to create as unsubscribe "buckets." I've organized each that we have in the dashboard currently into one of the below. We only have two groups for unsubscribes so far, with more to come likely in the future. The study activities are essential comms and participants cannot unsubscribe from those. When the functionality is available, we will have to develop a message that pops up when someone tries to unsubscribe to let them know those comms cannot be unsubscribed from, but they can update their contact info preferred email and/or contact the CSC.

Let me know if you have nay questions:

1. Newsletter communications/study updates, acknowledgement emails: _can_ unsubscribe from these communcations:
category name: "newsletter"

2. Study activities: _cannot_ unsubscribe from these communications:
category names:
"Baseline Menstrual Cycle Survey Reminders"
"Baseline Research Biospecimen Survey Reminders"
"Cannot Be Verified Notification"
"Data Destruction Reminders"
"HF Baseline Biospecimen Scheduling Invitation"
"HIPAA revocation reminders"
"HP  Baseline Research Biospecimen Scheduling Invitation"
"KPCO baseline biospecimen invitations"
"KPGA baseline biospecimen invitations"
"KPHI baseline biospecimen invitations"
"KPNW baseline biospecimen invitations"
"UC Baseline Biospecimen Scheduling Invitation"
"Verified, BL Survey Reminders"
"Verified, BL Survey Reminders Backlog Pts 2nd contact"
"Verified, BL Survey Reminders Backlog Pts 3rd contact"
"Verified, BL Survey Reminders Backlog Pts 4th contact"
"Verified, BL Survey Reminders Backlog Pts 5th contact"
"Verified, BL Survey Reminders Backlog Pts 6th contact"
"Verified, BL Survey Reminders New Pts 2nd contact"
"Verified, BL Survey Reminders New Pts 3rd contact"
"Verified, BL Survey Reminders New Pts 4th contact"
"Verified, BL Survey Reminders New Pts 5th contact"
"Verified, BL Survey Reminders New Pts 6th contact"
"Withdrawal Reminders"
"post-consent user profile reminders"

**Comment by jhflorey** [comment]
Hi @Davinkjohnson , @depietrodeanna 
I tried something for this issue
- 	Perform some operations on sengrid (create unsubscribe group, send dynamic email with unsubscribe group).
- 	Read the connectFaas code
- 	Perform some operations on notifications management in SMDB
- 	Read the sendgrid api documentation https://docs.sendgrid.com/api-reference/how-to-use-the-sendgrid-v3-api/authentication

I have some ideas to use sendgrid api in connectFaas and also update something in SMDB.
- Added `allowed to unsubscribe` option. When this option is selected we will use sendgrid api https://docs.sendgrid.com/api-reference/suppressions-unsubscribe-groups/create-a-new-suppression-group to create a new unsubscribe group with the category name (if it doesn't already exist). Alternatively, we can create an unsubscribe group on sendgrid manually but this seems inflexible.
![image](https://github.com/episphere/connect/assets/24559107/c663028d-3422-4e0b-a107-a0605007c8f0)

-  We will then get the unsubscribe group id and update the data for the `notification schema`.
![image](https://github.com/episphere/connect/assets/24559107/8d8f5037-efd2-443a-90dc-94e39794f0ca)

- When the system runs the scheduled `sendEmailNotification` job, we check to see if the notification is allowed to unsubscribe. If yes, we will use sendgrid api https://docs.sendgrid.com/api-reference/mail-send/mail-send to add an unsubscribe group to the email being sent. Otherwise we will send normally

Another issue that concerns me is `When the functionality is available, we will have to develop a message that pops up when someone tries to unsubscribe to let them know those comms cannot be unsubscribed from, but they can update their contact info preferred email and/or contact the CSC.` => I do not know how to show the pops up for particiapnt, or we have to implement the unsubscription management for participant in PWA or SMDB.

Please let me know you thoughts.

**Comment by Davinkjohnson** [comment]
Jessica, I think the plan you outlined seems like it should meet the requirements as defined and be pretty scalable. I'd ask that you also review it from a technical perspective with one of the other Sr developers, since it's a new area of work for us.

As for your last concern, around the pop up on unsubscribe, can you run a test where you send an email to a participant that does not have "allow unsubscribe" and try to unsubscribe anyway?

I'm thinking that SendGrid manages the unsubscribe page and we'll therefore not have much control over it, but it would be best to know for sure.

**Comment by jhflorey** [comment]
@anthonypetersen @we-ai I know you are an expert in this field, so could you please let me know your thoughts and advice.

**Comment by we-ai** [comment]
Based on my understanding, Sendgrid unsubcription group is not a good solution:
- Participants can change emails, then emails will go through Sendgrid unblocked. We can track email changes and consistently maintain an updated email list, but this is some extra work and is error prone that we better avoid if possible.
- Notification Specifications are for email formats and schedules.  We should avoid adding and maintaining other data in them.

My suggestion is that we should maintain unsubscription info in participants table:
- When a participant choose to unsubscribe a letter category or even a specific email type, update data in the participant.
- When the scheduled emails are run, participants are filtered based on existing email spec conditions PLUS unsubcription info. Participants with unsubscription on the email should be skipped. This can be done in one SQL query, so it'll be fast.

**Comment by jhflorey** [comment]
@we-ai `When a participant choose to unsubscribe a letter category or even a specific email type, update data in the participant.` => it means we will add to the bottom of email the unsubscribe link (connectFaas api) to update the data of this participant?

**Comment by we-ai** [comment]
I'm not sure suggestions from Opts team about where to put the un-subscrition link/button. Adding the link to the survey UIs will be easier.
As an example, related structure in participants data can be:
```json
"query.notificationsUnsubscribed": {
"newsletter": true, 
"a1a482cd-2b13-4d6b-a1fd-4d397599d565": true,
}
```

**Comment by jhflorey** [comment]
Yes, I think we can put the link at the end of the email body.

**Comment by anthonypetersen** [comment]
Thoughts...

If SendGrid has functionality to maintain unsubscribe information for us, I think it's better we try to use that versus building / maintaining logic in our code base assuming it's not inconvenient to integrate with SendGrid. I'm not opposed to adding a field in the `notificationSpecifications` table about allowing to unsubscribe if that can actually drive logic within SendGrid. I would prefer that we send SendGrid an unsubscribe group name based on details about the notification that we already have. Warren brings up a good point though about changing emails. If SendGrid has an API where we can see what unsubscribe groups an email is apart of as well as add emails to an unsubscribe groups, we could easily add logic to our change email functionality where the new email would get added to the same unsubscribe groups as the old email.

In general, my thoughts on the subject can be best summarized as "Let's not try to re-invent the wheel if we don't have to. If possible we should try to rely on services we already use that are completely focused on solving this problem as opposed to trying to solve it ourselves."

**Comment by jhflorey** [comment]
> Jessica, I think the plan you outlined seems like it should meet the requirements as defined and be pretty scalable. I'd ask that you also review it from a technical perspective with one of the other Sr developers, since it's a new area of work for us.
> 
> As for your last concern, around the pop up on unsubscribe, can you run a test where you send an email to a participant that does not have "allow unsubscribe" and try to unsubscribe anyway?
> 
> I'm thinking that SendGrid manages the unsubscribe page and we'll therefore not have much control over it, but it would be best to know for sure.

Hi @Davinkjohnson, i noticed that Sendgrid  also supports [Custom Unsubscribe Links](https://docs.sendgrid.com/ui/sending-email/create-and-manage-unsubscribe-groups#using-a-custom-unsubscribe-link
) where participants click on the link they will open our page for updates their informations. I believe it meets with "Study Activities Communications" which do not allow unsubscribes but have the option for participants to update their contact information as Deanna metioned `When the functionality is available, we will have to develop a message that pops up when someone tries to unsubscribe to let them know those comms cannot be unsubscribed from, but they can update their contact info preferred email and/or contact the CSC.`

**Comment by jhflorey** [comment]
Hi @Davinkjohnson @depietrodeanna @anthonypetersen @we-ai.  Here is [the proposal](https://nih.app.box.com/integrations/officeonline/openOfficeOnline?fileId=1324474977228&sharedAccessCode=) for this issue . Please review it and give me your thoughts.

**Comment by jhflorey** [comment]
Please use this link https://nih.app.box.com/file/1324474977228

**Comment by sonyekere** [comment]
@jhflorey, can this issue be closed?

**Comment by robertsamm** [comment]
@depietrodeanna @HanaShiho do you know whether this issue can be closed or if it is still needed? If still needed, can you assign it a milestone?

**Comment by depietrodeanna** [comment]
This is high priority per the comms wishlist https://nih.app.box.com/integrations/officeonline/openOfficeOnline?fileId=1782449370141&sharedAccessCode=. Need to work with DevOps to develop an appropriate timeline for implementation and next steps. @Davinkjohnson
